### PR TITLE
Kmeans Clustering Data Recipe

### PR DESCRIPTION
**Pull request by shivam5992**
Data Recipe to perform KMeans Clustering on a dataset

Users can define (optional) user inputs: number of clusters, columns to ignore, columns to include. 
If not defined anything then the number of clusters is automatically optimized and all columns are included. 

it creates an extra column as the output in the dataset with cluster_id of each row. 

Screenshot 2020-01-10 at 4 01 03 PM

Originally authored by @goldentom42, modified by @shivam5992 : added user-inputs, validations, missing_values check

### CODE REVIEW

File: data/KMeansClustering.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+			features = include_columns
+		
+		## list to ignore specific columns given by user
+		features = [_f for _f in features if _f not in ignore_columns]
+
+		## handle columns with missing values 
+		ignore_ = [] 
+		X_df = X.to_pandas()
+		for col in features:
+			if X_df[col].dtype == "object": # ignore categorical columns
```
Comment (goldentom42) [inline data/KMeansClustering.py]: What I would do here is go simple label encoding for string categoricals.
```
    X_df[f"{col}_enc"] = LabelEncoder().fit_transform(X_df[col].to_numpy())
    del X_df[:, col]
```
No tested but should work, not ideal but at least rows sharing the same values are recognized by Kmeans.

Comment (shivam5992) [reply data/KMeansClustering.py]: I tested that same logic for categorical variables this morning. It was working well. But later removed it as many people online have suggested that there might be problems with the label encoded representation. Different representation may result in different clusters. 

Additionally, a variant of K-means ie. KModes might be a good choice. 

https://datascience.stackexchange.com/questions/22/k-means-clustering-for-mixed-numeric-and-categorical-data

File: data/KMeansClustering.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+		X_clust = MinMaxScaler().fit_transform(X_clust)
+
+		# Go through possible numbers of clusters
+		best_score = None
+		best_n_clust = None
+		best_clust_ids = None
+
+		## if number of clusters is pre-defined by user, then dont find the optimal
+		if num_clusters > 1:
+			n_clusters = num_clusters
```
[the remark this answers is not in the document]
Comment (shivam5992) [reply data/KMeansClustering.py]: Agreed, will fix.

File: data/KMeansClustering.py (hunk 1 above)
Comment (goldentom42) [reply data/KMeansClustering.py]: Sure this is not an ideal representation but it's something.

I think we could work on a PR in closed recipes to find better ways for categoricals and have LabelEncoding for now.

Comment (goldentom42) [reply data/KMeansClustering.py]: I never tried sparse representation for Kmeans and maybe H2O-3 has better support for that as well.

Comment (shivam5992) [reply data/KMeansClustering.py]: Agreed, Added the label encoding for categorical variables.
